go/gossip/storage.go

```diff
@@ -0,0 +1,339 @@
+		return err
+	}
+	if s.selectRandomRecentPollination, err = s.db.Prepare(selectRandomRecentPollination); err != nil {
+		return err
+	}
+	if s.selectSCTID, err = s.db.Prepare(selectSCTID); err != nil {
+		return err
+	}
+	if s.selectNumChains, err = s.db.Prepare(selectNumChains); err != nil {
+		return err
+	}
+	if s.selectNumFeedback, err = s.db.Prepare(selectNumFeedback); err != nil {
+		return err
+	}
+	if s.selectNumSCTs, err = s.db.Prepare(selectNumSCTs); err != nil {
+		return err
+	}
+	if s.selectNumSTHs, err = s.db.Prepare(selectNumSTHs); err != nil {
+		return err
+	}
+	if s.selectFeedback, err = s.db.Prepare(selectFeedback); err != nil {
+		return err
+	}
+	if s.selectSTH, err = s.db.Prepare(selectSTH); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close closes the underlying DB storage.
+func (s *Storage) Close() error {
+	return s.db.Close()
+}
+
+func selectThingID(getID *sql.Stmt, thing interface{}) (int64, error) {
+	rows, err := getID.Query(thing)
+	if err != nil {
+		return -1, err
+	}
+	if !rows.Next() {
+		return -1, fmt.Errorf("couldn't look up ID for %v", thing)
+	}
+	var id int64
+	if err = rows.Scan(&id); err != nil {
+		return -1, err
+	}
+	return id, nil
+}
+
+// insertThingOrSelectID will attempt to execute the insert Statement (under transaction tx), if that fails due to
+// a unique primary key constraint, it will look up that primary key by executing the getID Statement.
+// Returns the ID associated with persistent thing, or an error describing the failure.
+func insertThingOrSelectID(tx *sql.Tx, insert *sql.Stmt, getID *sql.Stmt, thing interface{}) (int64, error) {
+	txInsert := tx.Stmt(insert)
+	txGetID := tx.Stmt(getID)
+	r, err := txInsert.Exec(thing)
+	if err != nil {
+		switch err.(type) {
+		case sqlite3.Error:
+			if err.(sqlite3.Error).Code == sqlite3.ErrConstraint {
+				return selectThingID(txGetID, thing)
+			}
+		}
+		return -1, err
+	}
+	id, err := r.LastInsertId()
+	if err != nil {
+		return -1, err
+	}
+	return id, nil
+}
+
+func (s *Storage) addChainIfNotExists(tx *sql.Tx, chain []string) (int64, error) {
+	flatChain := strings.Join(chain, "")
+	return insertThingOrSelectID(tx, s.insertChain, s.selectChainID, flatChain)
+}
+
+func (s *Storage) addSCTIfNotExists(tx *sql.Tx, sct string) (int64, error) {
+	return insertThingOrSelectID(tx, s.insertSCT, s.selectSCTID, sct)
+}
+
+func (s *Storage) addSCTFeedbackIfNotExists(tx *sql.Tx, chainID, sctID int64) error {
+	stmt := tx.Stmt(s.insertSCTFeedback)
+	_, err := stmt.Exec(chainID, sctID)
+	if err != nil {
+		switch err.(type) {
+		case sqlite3.Error:
+			// If this is a dupe that's fine, no need to return an error
+			if err.(sqlite3.Error).Code != sqlite3.ErrConstraint {
+				return err
+			}
+		default:
+			return err
+		}
+	}
+	return nil
+}
+
+// AddSCTFeedback stores the passed in feedback object.
+func (s *Storage) AddSCTFeedback(feedback SCTFeedback) error {
+	tx, err := s.db.Begin()
```

Comment:
I guess it doesn't really need to be at the moment, but if we ever allowed the server to ditch old feedback/pollen then we'd potentially have a race here.